core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java

```diff
@@ -1163,6 +1163,52 @@ private static RelDataType mapReturnType(SqlOperatorBinding opBinding) {
         false);
   }
 
+  @SuppressWarnings("argument.type.incompatible")
+  private static RelDataType mapKeyReturnType(SqlOperatorBinding opBinding) {
+    List<RelDataType> operandType = new ArrayList<>();
+
+    operandType.add(opBinding.collectOperandTypes().get(0).getKeyType());
+    operandType.add(opBinding.collectOperandTypes().get(0).getValueType());
+    Pair<@Nullable RelDataType, @Nullable RelDataType> type =
+        getComponentTypes(
+            opBinding.getTypeFactory(), operandType);
+
+    requireNonNull(type.left, () -> "type left of " + type.left);
+    requireNonNull(type.right, () -> "type right of " + type.right);
+    if (type.left.getSqlTypeName() != SqlTypeName.UNKNOWN
+        && type.right.getSqlTypeName() != SqlTypeName.UNKNOWN) {
+      SqlValidatorUtil.adjustTypeForMapFunctionConstructor(type, opBinding);
```

Comment:
Is this function allowed to mutate the operators?
Can you point to another example which does that?
I would not expect type inference to modify the program.

Reply:
I think what you mean is whether there are other functions like map_values and map_key functions. I will test it later to determine whether there are similar functions.

Reply:
map_contains_keys?
```bash
scala> val df = spark.sql("select map_contains_key(map(1, 'a', 2, 'b'), 1.0)")
val df: org.apache.spark.sql.DataFrame = [map_contains_key(map(1, a, 2, b), 1.0): boolean]

scala> df.show()
+--------------------------------------+
|map_contains_key(map(1, a, 2, b), 1.0)|
+--------------------------------------+
|                                  true|
+--------------------------------------+
```

Reply:
I think I should modify this jira description, it seems that not only these two functions have problems.
We can listen to other people's opinions, what do you think?
@tanclary

Reply:
I am asking whether the returnTypeInference function is allowed to change the opBinding.
Did you see other examples in the code where this happens?
Neither map_concat or map_entries do that.

Reply:
Sorry, I misunderstood what you meant. Of course there are other similar encodings
https://issues.apache.org/jira/browse/CALCITE-5976
https://issues.apache.org/jira/browse/CALCITE-5948